witchcraft/internal/tcpjson/conn_provider.go

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tcpjson
+
+import (
+	"crypto/tls"
+	"net"
+	"net/url"
+	"sync/atomic"
+
+	werror "github.com/palantir/witchcraft-go-error"
+)
+
+// ConnProvider defines the behavior to retrieve an established net.Conn.
+type ConnProvider interface {
+	// GetConn returns a net.Conn or an error if there is no connection established.
+	// It is the callers responsibility to close the returned net.Conn and
+	// gracefully handle any closed connection errors if the net.Conn is shared across clients.
+	GetConn() (net.Conn, error)
+}
+
+const (
+	ErrNoURIs           = "no URIs provided to connect to"
+	ErrFailedParsingURI = "failed to parse provided URI"
+	ErrFailedDial       = "failed to dial the host:port"
+)
+
+var _ ConnProvider = (*tcpConnProvider)(nil)
+
+// tcpConnProvider implements a ConnProvider that will round-robin TCP connections to a specific set of hosts.
+type tcpConnProvider struct {
+	// nextHostIdx contains the index of the next host to connect to from the hosts slice below.
+	// The index will be reset to 0 when len(hosts) is reached to facilitate round-robin connections.
+	nextHostIdx int32
+	hosts       []string
+	tlsConfig   *tls.Config
+}
+
+func NewTCPConnProvider(uris []string, tlsCfg *tls.Config) (ConnProvider, error) {
```

Comment:
Add documentation for function that describes that:
* Returns a new `ConnProvider`
* Provided set of uris must have at least one element and must be able to be parsed as URIs
* If config is `nil`, then default is used